design-proposals/klip-47-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,122 @@
+Secondly, when calling `CREATE TABLE`, the actual data is neither materialized nor 
+ready for interactive queries to access. To address these gaps, we would like to add:
+
+* Syntax to add read-only stream/table as `CREATE SOURCE STREAM/TABLE` to make the data source as read-only. Any 
+insertion command will be rejected, with an example error response as:
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will materialize the input source topic into a stream 
+instance which would generate a RocksDB instance for interactive queries.
+
+
+## What is in scope
+* Add `CREATE SOURCE STREAM/TABLE` syntax to KSQL
+* `DESCRIBE ... EXTENDED` should show source streams and tables as read-only  
+* Add materialization of the `source table` to make it pull-query accessible
+* Make source stream/table immutable from KSQL by rejecting insertion requests
+ 
+
+## What is not in scope
+* We will not add role-based write permission for source stream/table in v1, even though it seems to be useful.
+
+## Design
+
+### Semantic Change
+We shall first add optional keyword `SOURCE` into the KSQL codebase, by adding the new syntax to `sqlBase.g4` for:
+* createStream
+* createStreamAs
+* createTable
+* createTableAs
+
+Then inside `AstBuilder#visitCreateStream` and `AstBuilder#visitCreateTable`, the built classes for stream/table 
+instances would be `CreateStream` and `CreateTable`, which are the only two subclasses of `CreateSource`. It makes sense 
+to add the `SOURCE` as a flag into `CreateSource`, and let `AstBuilder` pass the flag into the struct 
+for later stage reference of read-only.
+
+In addition, when user calls `DESCRIBE ... EXTENDED`, the read-only attribute shall be displayed for stream/table. 
+
+### Make Source Stream/Table Read-only
+We plan to add restriction check into the `DistributingExecutor#throwIfInsertOnReadOnlyTopic` or create a similar function, 
+which could verify whether the given stream/table is read-only. If so, the insertion will be rejected.
+
+### Source Table Materialization
+In the call `EngineExecutor#execute`, we will check whether given plan has a `queryPlan` before deciding to proceed 
+creating a persistent query. To bypass this check, we need to provide a concrete query plan within the input plan. We would 
+inject a dummy query plan in `EngineExecutor#plan` phase, when the statement is a CreateTable statement with `SOURCE` tag.
+
+In addition, when user tries to show all the running queries, the materialization query should also be displayed as it takes 
+part of the computation resource. This would be a special type of running query without `SELECT` clause.
+
+Here comes a precaution we might be aware. The original `CREATE TABLE` command does not need materialization, so there is 
+no enforcement for primary key. However, in order to let pull query run on the materialized state, the store must be keyed so that 
```

Comment:
> so there is no enforcement for primary key

From my understanding defining a PK is enforced for existing `CT` statement.

Reply:
Follow up question for @agavra @vcrfxia: How strict are we about verifying the physical schema of the key compare to the defined PK? For values, I know that we effectively just ignore additional data (ie, if the AVRO record has 10 fields, and a user only does a `CS` picking 4 field it is fine), and for missing fields we just put `NULL`.

However, for keys it seems we might break co-partitioning if we are not strict about it? Would we drop all records on read if the define PK does not exactly match the physical schema of the message key?

Reply:
No, we don't do any such verification today. Keys and values are deserialized in the same way.

Why is co-partitioning a concern with pull queries? The CREATE SOURCE TABLE statement is going to have to copy the data into a new state store to enable pull queries anyway. As long as ksqlDB controls the key serializer used for converting the record into bytes for the state store, then finding the proper partition at the time of issue a pull query for a specific key will not be an issue. 

OTOH, if the proposal is to directly copy bytes from the Kafka key and record into RocksDB, then serialization discrepancies at the time of pull queries being issued is indeed a concern. That would be a concern not only if the specified key is only a subset of the actual key, but also for Schema-Registry-enabled key formats in case the schema ID is different.

Reply:
I don't know this off the top of my head - but we use the same serde for keys/values so I suspect we're not strict about it. The only difference is that we pass in `isKey=true` to the (connect) serdes so that might change the way the handle it but I think it's unlikely.

We probably do want to be strict here.

Reply:
I think, without a repartitioning step, co-partitioning cannot be guaranteed, and it sound like as if ksqlDB would no re-partition? -- I though that if we run a source table program, we would just read from the topic and put into RocksDB...

Also for right now, when we run a persistent query, would we repartition? Assume you do a table-table join, do we repartition both input topic before the join?